HPCC-18153 Merge DESDL functionality into ESP itself

mayx opened the pull request:
- Create a centralized subscription/monitoring mechanism that would handle
  changes for all esdl bindings and definitions. (HPCC-18754)
- Be able to add esdl binding without requiring a static binding (HPCC-17860)
- Support hybrid mode, methods of a static esp service and those of
  a dynamic esdl service in one service (HPCC-18123)
- Be able to bind multiple esdl services on one port (HPCC-17379)
- Improve binding manipulation methods to solely use bindingId. (HPCC-18126)
- Other command line fixes and improvements (HPCC-18733)
- ESDL Engine should consider sharing ESDL definition objects (HPCC-13057)
- ESDL Publish command should not require ESDL service name (HPCC-18133)
- Decoupling of definition name and ESDL service name.

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

mayx commented:
@afishbeck @rpastrana Please review...

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 11e29bcf76bbed1d5ad2a1318e64d38bbf60d74b
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting mydali ...           \x1b[32m [   OK    ]  \x1b[0m 
Starting mydfuserver ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclagent ...       \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclccserver ...    \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclscheduler ...   \x1b[32m [   OK    ]  \x1b[0m 
Starting myesp ...            \x1b[33m [ TIMEOUT ] \x1b[0m 
Starting myroxie ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mysasha ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mythor ...           \x1b[32m [   OK    ]  \x1b[0m 
 
HPCC status:
mydafilesrv     ( pid    28552 ) is running ...
mydali          ( pid    28875 ) is running ...
mydfuserver     ( pid    29217 ) is running ...
myeclagent      ( pid    29566 ) is running ...
myeclccserver   ( pid    29914 ) is running ...
myeclscheduler  ( pid    30255 ) is running ...
myesp           is stopped,
myroxie         ( pid    31817 ) is running ...
mysasha         ( pid    32225 ) is running ...
mythor          ( pid      340 ) is running with 1 slave process(es) ...
myesp           is stopped,
Archive HPCC logs into /mnt/disk1/home/vamosax/smoketest/PR-11045/HPCCSystems-logs-2018-04-18_16-41-45.
Removing RPM
Loaded plugins: fastestmirror
Resolving Dependencies
--> Running transaction check
---> Package hpccsystems-platform.x86_64 0:6.5.0-trunk0 will be erased
--> Finished Dependency Resolution
Dependencies Resolved
================================================================================
 Package                   Arch        Version             Repository      Size
================================================================================
Removing:
 hpccsystems-platform      x86_64      6.5.0-trunk0        installed      168 M
Transaction Summary
================================================================================
Remove  1 Package
Installed size: 168 M
Downloading packages:
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
Stopping mythor...            \x1b[32m [   OK    ]  \x1b[0m 
Stopping mysasha...           \x1b[32m [   OK    ]  \x1b[0m 
Stopping myroxie...           \x1b[32m [   OK    ]  \x1b[0m 
Stopping myesp...             /opt/HPCCSystems/bin/start-stop-daemon: warning: failed to kill 30531: No such process
1 pids were not killed
No process in pidfile /var/run/HPCCSystems/init_myesp.pid found running; none killed.
\x1b[32m [   OK    ]  \x1b[0m 
Stopping myeclscheduler...    \x1b[32m [   OK    ]  \x1b[0m 
Stopping myeclccserver...     \x1b[32m [   OK    ]  \x1b[0m 
Stopping myeclagent...        \x1b[32m [   OK    ]  \x1b[0m 
Stopping mydfuserver...       \x1b[32m [   OK    ]  \x1b[0m 
Stopping mydali...            \x1b[32m [   OK    ]  \x1b[0m 
Service dafilesrv, mydafilesrv is still running.
To stop it, run service dafilesrv stop.
Stopping mydafilesrv...       \x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /etc/init.d/dafilesrv                        ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/dfuplus                             ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/ecl                                 ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/eclcc                               ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/eclplus                             ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/soapplus                            ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/testsocket                          ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/wuget                               ...\x1b[32m [   OK    ]  \x1b[0m 
Removi
 ... ( comment is too long, 115 lines, 6429 bytes, truncated) ... 
n```

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 29fefae9bd05d5891b4d423d8f477b2b46155252
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting mydali ...           \x1b[32m [   OK    ]  \x1b[0m 
Starting mydfuserver ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclagent ...       \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclccserver ...    \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclscheduler ...   \x1b[32m [   OK    ]  \x1b[0m 
Starting myesp ...            \x1b[33m [ TIMEOUT ] \x1b[0m 
Starting myroxie ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mysasha ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mythor ...           \x1b[32m [   OK    ]  \x1b[0m 
 
HPCC status:
mydafilesrv     ( pid    16301 ) is running ...
mydali          ( pid    16622 ) is running ...
mydfuserver     ( pid    16971 ) is running ...
myeclagent      ( pid    17318 ) is running ...
myeclccserver   ( pid    17654 ) is running ...
myeclscheduler  ( pid    17996 ) is running ...
myesp           is stopped,
myroxie         ( pid    19622 ) is running ...
mysasha         ( pid    20030 ) is running ...
mythor          ( pid    20609 ) is running with 1 slave process(es) ...
myesp           is stopped,
Archive HPCC logs into /mnt/disk1/home/vamosax/smoketest/PR-11045/HPCCSystems-logs-2018-04-18_17-55-21.
Removing RPM
Loaded plugins: fastestmirror
Resolving Dependencies
--> Running transaction check
---> Package hpccsystems-platform.x86_64 0:6.5.0-trunk0 will be erased
--> Finished Dependency Resolution
Dependencies Resolved
================================================================================
 Package                   Arch        Version             Repository      Size
================================================================================
Removing:
 hpccsystems-platform      x86_64      6.5.0-trunk0        installed      168 M
Transaction Summary
================================================================================
Remove  1 Package
Installed size: 168 M
Downloading packages:
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
Stopping mythor...            \x1b[32m [   OK    ]  \x1b[0m 
Stopping mysasha...           \x1b[32m [   OK    ]  \x1b[0m 
Stopping myroxie...           \x1b[32m [   OK    ]  \x1b[0m 
Stopping myesp...             /opt/HPCCSystems/bin/start-stop-daemon: warning: failed to kill 18273: No such process
1 pids were not killed
No process in pidfile /var/run/HPCCSystems/init_myesp.pid found running; none killed.
\x1b[32m [   OK    ]  \x1b[0m 
Stopping myeclscheduler...    \x1b[32m [   OK    ]  \x1b[0m 
Stopping myeclccserver...     \x1b[32m [   OK    ]  \x1b[0m 
Stopping myeclagent...        \x1b[32m [   OK    ]  \x1b[0m 
Stopping mydfuserver...       \x1b[32m [   OK    ]  \x1b[0m 
Stopping mydali...            \x1b[32m [   OK    ]  \x1b[0m 
Service dafilesrv, mydafilesrv is still running.
To stop it, run service dafilesrv stop.
Stopping mydafilesrv...       \x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /etc/init.d/dafilesrv                        ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/dfuplus                             ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/ecl                                 ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/eclcc                               ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/eclplus                             ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/soapplus                            ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/testsocket                          ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/wuget                               ...\x1b[32m [   OK    ]  \x1b[0m 
Removi
 ... ( comment is too long, 115 lines, 6429 bytes, truncated) ... 
n```

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: Ubuntu 16.04 (Linux 4.4.0-119-generic)
Sha: 29fefae9bd05d5891b4d423d8f477b2b46155252
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0xenial_amd64.deb
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting mydali ...           \x1b[32m [   OK    ]  \x1b[0m 
Starting mydfuserver ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclagent ...       \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclccserver ...    \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclscheduler ...   \x1b[32m [   OK    ]  \x1b[0m 
Starting myesp ...            \x1b[33m [ TIMEOUT ] \x1b[0m 
Starting myroxie ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mysasha ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mythor ...           \x1b[32m [   OK    ]  \x1b[0m 
 
HPCC status:
mydafilesrv     ( pid    20285 ) is running ...
mydali          ( pid    20551 ) is running ...
mydfuserver     ( pid    20836 ) is running ...
myeclagent      ( pid    21127 ) is running ...
myeclccserver   ( pid    21404 ) is running ...
myeclscheduler  ( pid    21705 ) is running ...
myesp           is stopped,
myroxie         ( pid    23181 ) is running ...
mysasha         ( pid    23521 ) is running ...
mythor          ( pid    23985 ) is running with 1 slave process(es) ...
myesp           is stopped,
Archive HPCC logs into /home/ati/smoketest/PR-11045/HPCCSystems-logs-2018-04-19_17-44-53.
Keep files therefore skip HPCC Uninstall
Remove Cassandra leftovers.
Finish at: 2018-04-19_18-06-07
time logs started
Prep time   :    51 sec (00:00:51)
Build time  :   855 sec (00:14:15)
Package time:    54 sec (00:00:54)
Install time:   167 sec (00:02:47)
Start time  :   313 sec (00:05:13)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :  1440 sec (00:24:00)
time logs end
All done.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: Ubuntu 16.04 (Linux 4.4.0-119-generic)
Sha: 29fefae9bd05d5891b4d423d8f477b2b46155252
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0xenial_amd64.deb
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...      [   OK    ] 
Starting mydali ...           [   OK    ] 
Starting mydfuserver ...       [   OK    ] 
Starting myeclagent ...        [   OK    ] 
Starting myeclccserver ...     [   OK    ] 
Starting myeclscheduler ...    [   OK    ] 
Starting myesp ...            [ TIMEOUT ] 
Starting myroxie ...           [   OK    ] 
Starting mysasha ...           [   OK    ] 
Starting mythor ...            [   OK    ] 
 
HPCC status:
mydafilesrv     ( pid    20285 ) is running ...
mydali          ( pid    20551 ) is running ...
mydfuserver     ( pid    20836 ) is running ...
myeclagent      ( pid    21127 ) is running ...
myeclccserver   ( pid    21404 ) is running ...
myeclscheduler  ( pid    21705 ) is running ...
myesp           is stopped,
myroxie         ( pid    23181 ) is running ...
mysasha         ( pid    23521 ) is running ...
mythor          ( pid    23985 ) is running with 1 slave process(es) ...
myesp           is stopped,
Archive HPCC logs into /home/ati/smoketest/PR-11045/HPCCSystems-logs-2018-04-19_17-44-53.
Keep files therefore skip HPCC Uninstall
Remove Cassandra leftovers.
Finish at: 2018-04-19_18-06-07
time logs started
Prep time   :    51 sec (00:00:51)
Build time  :   855 sec (00:14:15)
Package time:    54 sec (00:00:54)
Install time:   167 sec (00:02:47)
Start time  :   313 sec (00:05:13)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :  1440 sec (00:24:00)
time logs end
All done.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: Ubuntu 16.04 (Linux 4.4.0-119-generic)
Sha: 29fefae9bd05d5891b4d423d8f477b2b46155252
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0xenial_amd64.deb
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...     [   OK    ]  
Starting mydali ...          [   OK    ]  
Starting mydfuserver ...     [   OK    ]  
Starting myeclagent ...      [   OK    ]  
Starting myeclccserver ...   [   OK    ]  
Starting myeclscheduler ...  [   OK    ]  
Starting myesp ...           [ TIMEOUT ] 
Starting myroxie ...         [   OK    ]  
Starting mysasha ...         [   OK    ]  
Starting mythor ...          [   OK    ]  
 
HPCC status:
mydafilesrv     ( pid     7141 ) is running ...
mydali          ( pid     7421 ) is running ...
mydfuserver     ( pid     7720 ) is running ...
myeclagent      ( pid     8023 ) is running ...
myeclccserver   ( pid     8314 ) is running ...
myeclscheduler  ( pid     8629 ) is running ...
myesp           is stopped,
myroxie         ( pid    10107 ) is running ...
mysasha         ( pid    10461 ) is running ...
mythor          ( pid    10941 ) is running with 1 slave process(es) ...
myesp           is stopped,
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   21 sec (00:00:21) |  916 sec (00:15:16) |   55 sec (00:00:55) |  142 sec (00:02:22) |  287 sec (00:04:47) |    0 sec (00:00:00) |    0 sec (00:00:00) | 1421 sec (00:23:41) |

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 29fefae9bd05d5891b4d423d8f477b2b46155252
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...     [   OK    ]  
Starting mydali ...          [   OK    ]  
Starting mydfuserver ...     [   OK    ]  
Starting myeclagent ...      [   OK    ]  
Starting myeclccserver ...   [   OK    ]  
Starting myeclscheduler ...  [   OK    ]  
Starting myesp ...           [ TIMEOUT ] 
Starting myroxie ...         [   OK    ]  
Starting mysasha ...         [   OK    ]  
Starting mythor ...          [   OK    ]  
 
HPCC status:
mydafilesrv     ( pid    17730 ) is running ...
mydali          ( pid    18052 ) is running ...
mydfuserver     ( pid    18394 ) is running ...
myeclagent      ( pid    18738 ) is running ...
myeclccserver   ( pid    19073 ) is running ...
myeclscheduler  ( pid    19413 ) is running ...
myesp           is stopped,
myroxie         ( pid    20969 ) is running ...
mysasha         ( pid    21376 ) is running ...
mythor          ( pid    21956 ) is running with 1 slave process(es) ...
myesp           is stopped,
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    5 sec (00:00:05) |  184 sec (00:03:04) |   56 sec (00:00:56) |    6 sec (00:00:06) |  141 sec (00:02:21) |    0 sec (00:00:00) |    0 sec (00:00:00) |  392 sec (00:06:32) |

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 5d20750ec3e22786d1f9960bcf8c788247fa2a6a
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...     [   OK    ]  
Starting mydali ...          [   OK    ]  
Starting mydfuserver ...     [   OK    ]  
Starting myeclagent ...      [   OK    ]  
Starting myeclccserver ...   [   OK    ]  
Starting myeclscheduler ...  [   OK    ]  
Starting myesp ...           [ TIMEOUT ] 
Starting myroxie ...         [   OK    ]  
Starting mysasha ...         [   OK    ]  
Starting mythor ...          [   OK    ]  
 
HPCC status:
mydafilesrv     ( pid    20436 ) is running ...
mydali          ( pid    20758 ) is running ...
mydfuserver     ( pid    21098 ) is running ...
myeclagent      ( pid    21446 ) is running ...
myeclccserver   ( pid    21780 ) is running ...
myeclscheduler  ( pid    22119 ) is running ...
myesp           is stopped,
myroxie         ( pid    23673 ) is running ...
mysasha         ( pid    24081 ) is running ...
mythor          ( pid    24658 ) is running with 1 slave process(es) ...
myesp           is stopped,
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   19 sec (00:00:19) |  186 sec (00:03:06) |   55 sec (00:00:55) |    6 sec (00:00:06) |  142 sec (00:02:22) |    0 sec (00:00:00) |    0 sec (00:00:00) |  408 sec (00:06:48) |

rpastrana commented:
@mayx the smoketest seems to be failing

mayx commented:
@rpastrana yes looking into it. It seems the symbol exposure behaves slightly differently on my local build from the smoketest build...

rpastrana commented:
@mayx should the code review wait for your changes?

mayx commented:
Yes, I'm working with Attila to get the logs and see what went wrong...

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 4e65507f71d54730235a24681b233cf0279aa10c
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    6 sec (00:00:06) |  184 sec (00:03:04) |   56 sec (00:00:56) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1160 sec (00:19:20) |   17 sec (00:00:17) | 1450 sec (00:24:10) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e30ffc256c44bb43329c3bbbb9c7eef5e266df00
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   24 sec (00:00:24) |  183 sec (00:03:03) |   56 sec (00:00:56) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1142 sec (00:19:02) |   16 sec (00:00:16) | 1448 sec (00:24:08) |

rpastrana commented on esp/esdllib/esdl_def.cpp:
```diff
@@ -1558,28 +1556,45 @@ IEsdlDefObjectIterator* EsdlDefinition::getDependencies( const char* service, St
-            if ((flags & DEPFLAG_ECL_ONLY) && methodIter->query().getPropInt("ecl_hide"))
-                continue;
-            methodArray.append( methodIter->query() );
+            Owned<IEsdlDefMethodIterator> methodIter = serviceDef->getMethods();
```
1568-1574 is repeated a few lines away this should be made common

rpastrana commented on esp/platform/espcfg.ipp:
```diff
@@ -198,13 +198,16 @@ public:
         loadServices();
         loadProtocols();
         loadBindings();
+        if(useDali)
```
as a heads up, the dali detach feature will be merged soon, you'll have to be aware of the current attach mode

rpastrana commented on esp/platform/espcfg.ipp:
```diff
@@ -253,7 +259,20 @@ public:
         closeEnvironment();
     }
 
+    IPropertyTree* queryProcPT()
```
minor, but it might be better renamed to queryProcConfig

rpastrana commented on esp/platform/espp.hpp:
```diff
@@ -229,6 +229,47 @@ class CEspServer : public CInterface,
+            {
+                DBGLOG("No more bindings on port %d, so freeing up the port.",port);
+                ISocket **socketp = m_srvSockets.getValue(port);
+                ISocket *socket=(socketp!=NULL) ? *socketp : NULL;
```
nullptr

rpastrana commented on esp/platform/espp.hpp:
```diff
@@ -229,6 +229,47 @@ class CEspServer : public CInterface,
+        }
+    }
+
+    virtual IPropertyTree* queryProcPT()
```
same as before, queryProcConfig?

rpastrana reviewed (changes requested):
@mayx started review, left a couple of comments. got as far as CEspApplicationPort::removeBinding

rpastrana commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -670,27 +827,137 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
+    IPropertyTree * getEspBindingConfig(const char * espprocname, const char * espbindingport, const char * bindingname)
+    {
+        if (!espprocname || !*espprocname)
+            return NULL;
```
nullptr

rpastrana commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -670,27 +827,137 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
+        globalLock->close();
+
+        if (espbindingport && *espbindingport)
+            xpath.clear().appendf("EspProcess/[@name='%s']/EspBinding[@port='%s']", espprocname, espbindingport);
```
setf?

rpastrana commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -732,128 +994,118 @@ class CEsdlSDSSubscription : implements IEsdlSubscription, public CInterface, im
-        VStringBuffer fullBindingPath("/ESDL/Bindings/Binding[@id=\'%s.%s\']",mProcessName.str(), mBindingName.str());
         CriticalBlock b(daliSubscriptionCritSec);
-        if(binding_sub_id == 0)
+        if (sub_id == 0)
```
mentioned before, but this needs to be aware of the dali attach/subscribe state

rpastrana commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -732,128 +994,118 @@ class CEsdlSDSSubscription : implements IEsdlSubscription, public CInterface, im
             {
-                definition_sub_id = querySDS().subscribe(ESDL_DEF_PATH, *this, true);
-                DBGLOG("Esdl definition subscription to DALI (%s) succeeded.", ESDL_DEF_PATH);
+                if (action && *action)
```
consider lowercasing action and changing the stricmp to case sensitive compare

rpastrana commented on tools/esdlcmd/esdl-publish.cpp:
```diff
@@ -128,8 +128,8 @@ class EsdlPublishCmd : public EsdlPublishCmdCommon
         }
-
-        request->setServiceName(optESDLService.get());
+        if (optESDLService.length() > 0)
```
!=0 ?

rpastrana reviewed (changes requested):
@mayx great features. Left a few comments. you might encounter some conflicts due to the recent pull request regarding Dali Detach. Also please confirm the critsecs aren't protecting a wider scope than intended (some might need their own local scope.

mayx commented:
@rpastrana thanks for reminding me about the conflicts. I resolved the conflicts and checked in the changes. Will go through your review comments and make corresponding changes next.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 3f4a6cc0f56010cddbfb0667ad7a4c3f2a4cf943
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  182 sec (00:03:02) |   56 sec (00:00:56) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1176 sec (00:19:36) |   15 sec (00:00:15) | 1470 sec (00:24:30) |

rpastrana commented:
bouncing back to @mayx

mayx commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,549 @@
+        Owned<CEsdlInstance> instance = new CEsdlInstance();
+        instance->m_defId.set(defId);
+        instance->m_def.set(def);
+        CriticalBlock cb(m_critSect);
```
Don't think it matters here as the scope starts from when it's declared (not from beginning of the method) to the end fo the method.

mayx commented:
@rpastrana changes checked in based on your comments, please review again. The changes are in a separate commit.

rpastrana commented on esp/esdllib/esdl_def.cpp:
```diff
@@ -1558,28 +1556,45 @@ IEsdlDefObjectIterator* EsdlDefinition::getDependencies( const char* service, St
         throw( MakeStringException(0, "No ESDL Service Definition provided, need to have a service to search in for methods") );
     }
 
     if( methods.length() < 1 )
```
we might want to change this one to ==0

rpastrana commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -466,7 +466,7 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
 
         if (!isEsdlServiceDefined(definitionId, esdlServiceName))
         {
-            message.setf("Esdl service %s is not defined in %s, please double check the case sensitive service name", esdlServiceName, definitionId);
```
case-sensitive

rpastrana reviewed (changes requested):
@mayx looks good. just caught a couple of minor issues.
Can we test the detach mode? from espcontrol, request DETACH forced=true, shut down dali, after a minute or so bring dali back up, perform a DESDLConfig change and verify?

rpastrana commented:
Jirabot has me as owner of this pr, sending to @mayx for review

mayx commented:
@rpastrana I made the 2 minor changes and squashed the commits. Also I tested the the detach mode as you suggested and it worked. If everything looks OK please approve the PR...

rpastrana reviewed (approved):
@mayx looks good.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: d680b23e4418ffb5bcf6ba8d2e6c35630904cae0
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  185 sec (00:03:05) |   56 sec (00:00:56) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1164 sec (00:19:24) |   16 sec (00:00:16) | 1466 sec (00:24:26) |

richardkchapman commented on esp/bindings/http/platform/httpservice.cpp:
```diff
@@ -104,6 +104,7 @@ bool CEspHttpServer::rootAuth(IEspContext* ctx)
     EspHttpBinding* thebinding=getBinding();
     if (thebinding)
     {
+        Owned<IInterface> theBindingHolder = dynamic_cast<IInterface*>(thebinding);
```
Dynamic casts without checking reslut for null will give coverity warnings. Is there any reason this cast needs to be a dynamic one?

mayx replied:
Dynamic cast is necessary because it's casting sideways. I'll add the check for nullptr.

ghalliday replied:
I don't think you have added the check yet.

richardkchapman commented on esp/bindings/http/platform/httpservice.cpp:
```diff
@@ -331,25 +334,44 @@ int CEspHttpServer::processRequest()
             }
+
+            if (thebinding)
+                theBindingHolder.set(dynamic_cast<IInterface*>(thebinding));
```
Comment as above

richardkchapman commented on esp/platform/espprotocol.cpp:
```diff
@@ -90,6 +91,29 @@ void CEspApplicationPort::appendBinding(CEspBindingEntry* entry, bool isdefault)
+            Owned<CDelayedBindingRemover> remover = new CDelayedBindingRemover(bindings[i]);
+            remover->start();
+            bindings[i] = nullptr;
+            if (i != bindingCount-1)
```
This doesn't feel like a case where efficiency is more important than clarity...

richardkchapman commented on esp/platform/espprotocol.cpp:
```diff
@@ -90,6 +91,29 @@ void CEspApplicationPort::appendBinding(CEspBindingEntry* entry, bool isdefault)
+        IEspRpcBinding* currentBinding = bindings[i]->queryBinding();
+        if (currentBinding && currentBinding == binding)
+        {
+            Owned<CDelayedBindingRemover> remover = new CDelayedBindingRemover(bindings[i]);
```
This feels iffy. You are starting a thread that will have no owner - will it leak? Will it release itself when it completes? Why does it need releasing asynchronously anyway?

mayx replied:
The run method of the thread links at the beginning and releases at the end. But you're right there might be a chance the object will be released before the link happens. The asynchronous release was put in place out of paranoia that the binding could be released while requests are still being processed. But since a holder for the binding was added in the request processing method, this shouldn't be a concern. I'll just remove the thread.

richardkchapman replied:
You may want to arrange for the release to happen separately from the removal from the list - so that you don't hold onto the write lock any longer than you need to. But that does not require a separate thread.

richardkchapman commented:
@ghalliday I'd like a second opinion on this

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 7aaf8424cbe2c1c77c0833f4d128202e6017011d
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   30 sec (00:00:30) |  180 sec (00:03:00) |   56 sec (00:00:56) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1188 sec (00:19:48) |   17 sec (00:00:17) | 1498 sec (00:24:58) |

mayx committed:
HPCC-18153 Merge DESDL functionality into ESP itself

- Create a centralized subscription/monitoring mechanism that would handle
  changes for all esdl bindings and definitions. (HPCC-18754)
- Be able to add esdl binding without requiring a static binding (HPCC-17860)
- Support hybrid mode, methods of a static esp service and those of
  a dynamic esdl service in one service (HPCC-18123)
- Be able to bind multiple esdl services on one port (HPCC-17379)
- Improve binding manipulation methods to solely use bindingId. (HPCC-18126)
- Other command line fixes and improvements (HPCC-18733)
- ESDL Engine should consider sharing ESDL definition objects (HPCC-13057)
- ESDL Publish command should not require ESDL service name (HPCC-18133)
- Decoupling of definition name and ESDL service name.
- Changes needed for merging with recent dali detach and attach changes.
- De-clutter esdl store code by removing unnecessary closes and
  queryroot() nullptr checks
- Remove one unnecessary method from esdl store.
- Fix a reference count issue by replacing Owned With Linked
- Memory leak fixes
- More changes based on code review comments and Valgrind MemCheck output

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

mayx commented:
@richardkchapman changes made per discussion.

ghalliday commented:
@ghalliday

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: c9afffc62069f57a7f23438d754f2773b087284b
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   31 sec (00:00:31) |  183 sec (00:03:03) |   56 sec (00:00:56) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1166 sec (00:19:26) |   16 sec (00:00:16) | 1480 sec (00:24:40) |

ghalliday commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -530,7 +646,29 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
+        }
+        else
+        {
+            conn->close();
```
I want to check with @jakesmith if this is legal. (It occurs quite a lot in this file).  Especially since close() has this comment:
    virtual void close(bool deleteRoot=false) = 0; // invalidates root.

jakesmith replied:
It is not safe, should be changed.
If you're going to close the connection, you need to finish with it's contents 1st.

ghalliday replied:
@jakesmith is the createPTreeFromIPT correct - (I think it is but want to double check).  Again I don't think the close gains anything.

ghalliday commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -670,27 +826,137 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
+        }
+        catch (ISDSException * e)
+        {
+            if (conn)
```
picky: it is impossible for conn to be non null!

ghalliday replied:
i.e. no point in checking for non null and then closing, since it must be null.

ghalliday commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -116,6 +116,50 @@ IPropertyTree * fetchConfigInfo(const char * config,
+        StringBuffer rootname;
+        configTree->getName(rootname);
+
+        IPropertyTree * deftree = NULL;
```
picky: Moving the variables nearer to where they are used would make this easier to read/understand.  E.g., line 143.

ghalliday commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -1491,40 +1636,79 @@ bool CWsESDLConfigEx::onListDESDLEspBindings(IEspContext &context, IEspListDESDL
-bool CWsESDLConfigEx::onListESDLBindings(IEspContext &context, IEspListESDLBindingsRequest&req, IEspListESDLBindingsResponse &resp)
+static int bindingCompareFunc(IInterface * const *_itm1, IInterface * const* _itm2)
 {
+    IConstESDLBinding* itm1 = dynamic_cast<IConstESDLBinding*>(*_itm1);
```
Should really be static casts (more efficient and more accurate).  Any dynamic cast needs to check that the result is not null.

ghalliday commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -1491,40 +1636,79 @@ bool CWsESDLConfigEx::onListDESDLEspBindings(IEspContext &context, IEspListDESDL
+            {
+                currentProcess.setown(createEspProcessStruct());
+                currentProcess->setName(binding->getEspProcess());
+                processes.append(*dynamic_cast<IConstEspProcessStruct*>(currentProcess.getLink()));
```
This cast shouldn't be needed (IConstEspProcessStruct is a base class of IEspEspProcessStruct), and definitely should not be a dynamic cast.

ghalliday commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -1491,40 +1636,79 @@ bool CWsESDLConfigEx::onListDESDLEspBindings(IEspContext &context, IEspListDESDL
+                IArrayOf<IConstEspPortStruct>& ports = currentProcess->getPorts();
+                currentPort.setown(createEspPortStruct());
+                currentPort->setValue(binding->getPort());
+                ports.append(*dynamic_cast<IConstEspPortStruct*>(currentPort.getLink()));
```
Also not needed, and confusing.

ghalliday commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,579 @@
+#include "esdl_svc_engine.hpp"
+#include <memory>
+
+class CEsdlInstance : implements IInterface, public CInterface
```
minor: no need to derive from IInterface if not being used.

ghalliday commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,579 @@
+            CriticalBlock cb(m_critSect);
+            Owned<CEsdlInstance>* instancepp = m_esdlInstanceMap.getValue(defId);
+            if (instancepp)
+                instance = *instancepp;
```
minor.  Not sure why this isn't

    return (*instancepp)->m_def.get();

ghalliday commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,579 @@
+                    CEsdlInstance* instance = item.getValue();
+                    if (!instance || !instance->m_def.get())
+                        continue;
+                    CInterface* defptr = dynamic_cast<CInterface*>(instance->m_def.get());
```
IsShared (or an equivalent function) should be exposed in the IEsdlDefinition to avoid a dynamic cast.  It is exposing implementation details that could easily break.  E.g., if EspDefinition was changed to

    EsdlDefinition : public CInterfaceOf<IEsdlDefinition>

ghalliday commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,579 @@
+    Owned<IEsdlDefinition> m_def;
+};
+
+class CEsdlShare : implements IEsdlShare, public Thread
```
Should have comments describing what this class is here for.

ghalliday commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,579 @@
+    }
+
+    //Thread
+    virtual int run() override
```
Should have a comment describing what the purpose of this thread is..

ghalliday commented on esp/platform/espp.hpp:
```diff
@@ -232,6 +232,47 @@ class CEspServer : public CInterface,
+            if (left == 0)
+            {
+                DBGLOG("No more bindings on port %d, so freeing up the port.",port);
+                ISocket **socketp = m_srvSockets.getValue(port);
```
Is there a potential race with addBinding?  E.g., if an add occurs at the same time as the last remove the binding could be removed, a new one added, then the socket removed - so nothing will be listening to the port.

ghalliday reviewed (commented):
A few comments on particular lines.  A couple of general comments:
- there seem to be too many dynamic casts, some not needed, some should be static, and some exposing implemention details.  They should really only be used as a last resort.
- New classes and functions should have a comment describing their main purpose.
- Multiple small PRs that each implement one JIRA are easier to review than one big PR!

I have concerns about all the calls to conn->close(), but defer to @jakesmith  on whether that is safe or not.

ghalliday reviewed (commented):
not sure if these are submitted

ghalliday commented:
@mayx see comments.  Note some are collapsed and need expanding..

jakesmith commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -63,7 +64,7 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
            throw MakeStringException(-1, "Unable to connect to ESDL Service definition information in dali '%s'", xpath.str());
 
-        conn->close(false); //release lock right away
+        conn->close(); //release lock right away
```
Interacting with the connection (e.g. conn->queryRoot()) after it's closed is not safe.
I'm surprised the toXML doesn't crash actually.

mayx replied:
I'm also surprised that it actually worked. I guess there might've been some confusion about what the default "false" parameter implies. I'll move the calls to the end of the operations.

jakesmith replied:
It would be worth adding a comment to clarify on close.
close(true) means delete the node it was connected to.

jakesmith commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -108,10 +109,42 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
         {
-            //There shouldn't be multiple entries here, but if so, we'll use the first one
-            VStringBuffer xpath("%s[@id='%s.%s'][1]", ESDL_BINDING_ENTRY, espProcess, espStaticBinding);
+            Owned<IRemoteConnection> conn = querySDS().connect(ESDL_BINDINGS_ROOT_PATH, myProcessSession(), RTM_LOCK_READ, SDS_LOCK_TIMEOUT_DESDL);
```
Yes, a lot. It avoids lazy fetching intermediate nodes up to the end target.

mayx commented:
@ghalliday Hi Gavin I've made changes based on your code review. Please review again when you have a chance...

ghalliday commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -64,14 +64,14 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
         //There shouldn't be multiple entries here, but if so, we'll use the first one
         toXML(deftree, esxdl, 0, 0);
+
+        conn->close(); //release lock right away
```
Does this gain anything?  The lock will be released as soon as the conn object goes out of scope.

jakesmith replied:
No, and dangerous, but I see it's now been removed.

ghalliday commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -109,20 +109,20 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
 
-            if (conn->queryRoot()->hasProp(xpath))
-                return createPTreeFromIPT(conn->queryRoot()->queryPropTree(xpath));
+            if (conn->queryRoot())
```
@jakesmith can queryRoot() ever be non null?

jakesmith replied:
Not at the moment, but it really should be cleared by close().
Either way it is dangerous/undefined behaviour after close().

ghalliday reviewed (commented):
A few follow on comments, mainly questions for Jake.

mayx commented:
@ghalliday changes made per your latest comments, checked in as "commit 3". Please review...

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a40d7002081b9729be98adaadd02c3da133751b2
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |784 |784 |0 |
|test (thor) |715 |715 |0 |
|test (roxie) |828 |828 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   16 sec (00:00:16) |  178 sec (00:02:58) |   60 sec (00:01:00) |    7 sec (00:00:07) |   27 sec (00:00:27) | 1166 sec (00:19:26) |   19 sec (00:00:19) | 1473 sec (00:24:33) |

jakesmith commented:
> ... is the createPTreeFromIPT correct - (I think it is but want to double check). Again I don't think the close gains anything.

Yes correct call.

jakesmith commented on esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp:
```diff
@@ -151,7 +195,7 @@ IPropertyTree * CWsESDLConfigEx::getEspProcessRegistry(const char * espprocname,
         throw MakeStringException(-1, "Unable to connect to ESP configuration information in dali %s", xpath.str());
 
-    globalLock->close(false);
+    globalLock->close();
```
globalLock.clear() would be better

mayx replied:
The getEspProcessRegistry() method creates unnecessary overhead, I'll remove the method.

jakesmith commented on esp/bindings/http/platform/httpservice.cpp:
```diff
@@ -105,6 +105,7 @@ bool CEspHttpServer::rootAuth(IEspContext* ctx)
     EspHttpBinding* thebinding=getBinding();
     if (thebinding)
     {
+        Linked<IInterface> theBindingHolder = dynamic_cast<IInterface*>(thebinding);
```
this looks a bit odd/suspicious.
thebinding is of type EspHttpBinding, which does appear to be a IInterface

ghalliday replied:
There should also be a check that theBindingHolder != null.  If it should never occur then assertex(theBindingHolder)

mayx replied:
I found that the rootAuth method is no longer being used, I'll just remove it.

jakesmith commented on esp/bindings/http/platform/httpservice.cpp:
```diff
@@ -336,25 +339,44 @@ int CEspHttpServer::processRequest()
+                                {
+                                    thebinding=lbind;
+                                    StringBuffer bindSvcName;
+                                    if (stricmp(serviceName, lbind->getServiceName(bindSvcName)))
```
just checking - this isn't to be a not equal check ?
could use !streq for clarity.

jakesmith commented on esp/bindings/http/platform/httpservice.cpp:
```diff
@@ -336,25 +339,44 @@ int CEspHttpServer::processRequest()
+                                {
+                                    exactBinding = lbind;
+                                    StringBuffer bindSvcName;
+                                    if (stricmp(serviceName, lbind->getServiceName(bindSvcName)))
```
similar to above, if (!streq(..., ..)) would be clearer.

jakesmith commented on esp/esdllib/esdl_def.cpp:
```diff
@@ -39,6 +39,7 @@ typedef CopyReferenceArrayOf<IEsdlDefMethod> EsdlDefMethodArray;
     virtual const char *queryProp(const char *name) { return EsdlDefObject::queryProp(name); } \
 virtual bool hasProp(const char* name) { return EsdlDefObject::hasProp(name); } \
     virtual int getPropInt(const char *pname, int def=0){return props->getPropInt(pname,def);} \
+    virtual bool getPropBool(const char *pname, bool def=false){return props->getPropBool(pname,def);} \
```
trivial/not new: odd indentation on line 40

ghalliday commented:
@mayx commit #3 looks much cleaner.  I'll read it all through again tomorrow.  PR needs rebasing.

jakesmith commented on esp/esdllib/esdl_def.cpp:
```diff
@@ -1464,6 +1466,7 @@ class EsdlDefinition : public CInterface, implements IEsdlDefinition
     unsigned walkChildrenDepthFirst( AddedObjs& foundByName, EsdlDefObjectWrapperArray& dependencies, IEsdlDefObject* esdlObj, double requestedVer, IProperties *opts, int level=0, unsigned flags=0);
     bool shouldGenerateArrayOf( IEsdlDefObject& child, unsigned flags, unsigned& stateOut );
     bool setWrapperFlagStringArray( AddedObjs& foundByName, EsdlDefObjectWrapper* wrapper, unsigned state );
+    virtual bool isShared() override { return IsShared(); }
```
curious why need this and not just the underlying non-virtual existing IsShared() method?

mayx replied:
Because we want to avoid a dynamic_cast to CInterface*.

jakesmith commented on esp/esdllib/esdl_def.cpp:
```diff
@@ -1590,10 +1602,22 @@ IEsdlDefObjectIterator* EsdlDefinition::getDependencies( const char* service, St
+        if (serviceDef)
+        {
+            EsdlDefObjectWrapper* wrapper = new EsdlDefObjectWrapper( *serviceDef );
+            dependencies->append( *wrapper );
```
not directly related to this PR, but I notice 'dependencies' would leak (line 1556) if an exception was thrown. More uses of Owned / getClear() etc. would be good.

mayx replied:
EsdlDefObjectWrapperArray is not inherited from Interface so Owned won't work. I'll create a simple conditional releasing class to deal with this scenario.

mayx replied:
Actually jlib's OwnedPtr class is perfect for this scenario.

jakesmith commented on esp/platform/espcfg.cpp:
```diff
@@ -466,31 +466,33 @@ CEspConfig::CEspConfig(IProperties* inputs, IPropertyTree* envpt, IPropertyTree*
+                        ptree->getProp("@netAddress", addr);
+                        if (strcmp(addr.str(), ".") == 0)
+                        {
+                            bcfg->address.append("0.0.0.0");
```
could do with a comment to say why if address contains a '.' it becomes "0.0.0.0"

jakesmith commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,576 @@
+        }
+    }
+
+    virtual Linked<IEsdlDefinition> query(const char* defId) override
```
this looks a bit questionable, especially given method name is query.
Shouldn't it just be:
virtual IEsdlDefinition *query(const char* defId) override

mayx replied:
The purpose of returning Linked is to hold on to the object right away, because there's a background thread that could clean up the object at any instant. I'll change the "query" method name which is too suggestive to a more neutral name "lookup".

mayx replied:
The purpose for returning Linked is to hold on to the definition right away, because there's a cleaning up thread that may delete the definition at any instant. I agree that the "query" method name is suggestive, so I'll change it to more neutral "lookup".

jakesmith commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,576 @@
+        constructEnvptHeader();
+        m_pCentralStore.setown(createEsdlCentralStore());
+        m_esdlShare.setown(new CEsdlShare());
+        m_esdlShare->start();
```
What stops and/or joins this thread?

mayx replied:
The stopEsdlMonitor() method which will be called by ESP's shutdown mechanism will stop this thread.

jakesmith replied:
ok I see, it's quite confusing. It only works because there's 1 instance of CEsdlMonitor really.

Also, stopEsdlMonitor() calls stop() but doesn't join the thread, which could lead to problems.

It would be better to contain the thread in the owning class and have a EsdlMonitor::stop() method, stop and join it's own thread.

mayx replied:
I've added the join to the stop method. Good point to add the join, but I think logically it makes more sense to keep the thread in the share class instead of the monitor class.

jakesmith commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,576 @@
+
+    virtual ~CEsdlMonitor()
+    {
+        if (m_pSubscription)
```
the unsubscribe method sets m_isSubscribed=false.. it looks like this should check that before trying to unsubscribe again.

ghalliday replied:
It checks it in unsubscribe, but not in this destructor.

jakesmith replied:
yes, still looks a problem to me.

jakesmith commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,576 @@
+
+    void loadDynamicBindings()
+    {
+        IPropertyTree * esdlBindings = m_pCentralStore->getBindings();
```
getXXX normally implies it returns a linked item, so it looks like this is going to leak..

jakesmith commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,576 @@
+                DBGLOG("ESDL binding %s doesn't have port specified, skip", data->id.str());
+                continue;
+            }
+            data->port = atoi(portStr.str());
```
why not use cur.getPropInt("@port") ?

jakesmith commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,576 @@
+        {
+            CriticalBlock cb(m_CritSect);
+            m_esdlShare->remove(data->id.str());
+            for (auto sb:m_esdlBindingMap)
```
should sb be a reference i.e. for (auto &sb:m_esdlBindingMap) ?

jakesmith commented on esp/services/esdl_svc_engine/esdl_monitor.hpp:
```diff
@@ -0,0 +1,44 @@
+{
+    virtual void add(const char* defId, IEsdlDefinition* defobj) = 0;
+    virtual void remove(const char* defId) = 0;
+    virtual Linked<IEsdlDefinition> query(const char* defId) = 0;
```
this looks a bit suspect (see other comment in impl.)

jakesmith commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -108,46 +103,77 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
 
     virtual IPropertyTree* fetchBinding(const char* espProcess, const char* espStaticBinding) override
     {
+        VStringBuffer xpath("%s[@espprocess='%s'][@espbinding='%s'][1]", ESDL_BINDING_PATH, espProcess, espStaticBinding);
```
There's a leak on line 97
>IPropertyTree * deftree = esdlDefinitions->getPropTree(xpath);

jakesmith commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -670,27 +761,116 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
+        else
+            xpath.setf("EspProcess/[@name='%s']/EspBinding[@name='%s']", espprocname, bindingname);
+        Owned<IPropertyTreeIterator> iter = globalLock->queryRoot()->getElements(xpath.str());
+        IPropertyTree* result = nullptr;
```
unused

jakesmith commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -670,27 +761,116 @@ class CEsdlSDSStore : implements IEsdlStore, public CInterface
+        Owned<IRemoteConnection> software = querySDS().connect(xpath.str(), myProcessSession(), RTM_LOCK_READ, SDS_LOCK_TIMEOUT_DESDL);
+        if (!software)
+            return false;
+        xpath.clear().appendf("EspProcess[@name='%s']", espprocname);
```
It would be better to try to connect directly to this for efficiency.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: f69b90d5b81bd096cf9af809a551f825061177af
Number of Error(s): 21
/mnt/disk1/home/vamosax/smoketest/PR-11045/HPCC-Platform/esp/services/ws_esdlconfig/ws_esdlconfigservice.hpp:70:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token

[Build error:](https://gist.github.com/168cfea826aadb91ab779f79a4f6cfa7)
Build log [Build log](https://gist.github.com/HPCCSmoketest/168cfea826aadb91ab779f79a4f6cfa7#file-result-18-05-10-05-41-29-log)

mayx commented:
@jakesmith @ghalliday Changes made per comments, commits squashed, please review again. (by the way the merge conflict above has been resolved).

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 84b972557b0bb4c91ec069932c94b87e7a870b6b
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |784 |784 |0 |
|test (thor) |715 |715 |0 |
|test (roxie) |828 |828 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   18 sec (00:00:18) |  169 sec (00:02:49) |   60 sec (00:01:00) |    8 sec (00:00:08) |   29 sec (00:00:29) | 1196 sec (00:19:56) |   19 sec (00:00:19) | 1499 sec (00:24:59) |

ghalliday commented on esp/platform/espcfg.cpp:
```diff
@@ -466,31 +466,34 @@ CEspConfig::CEspConfig(IProperties* inputs, IPropertyTree* envpt, IPropertyTree*
-                    }
+                    bcfg->port = ptree->getPropInt("@port", 0);
+                    if (bcfg->port == 0)
+                        DBGLOG("Binding %s is configured with port 0, do not load it.", bcfg->name.str());
```
This will leak bcfg

ghalliday commented on esp/platform/espprotocol.cpp:
```diff
@@ -68,6 +68,7 @@ CEspApplicationPort::CEspApplicationPort(bool viewcfg, CEspProtocol* prot) : bin
 void CEspApplicationPort::appendBinding(CEspBindingEntry* entry, bool isdefault)
 {
+    WriteLockBlock wblock(rwLock);
     bindings[bindingCount]=entry;
```
There should be a check that bindingCount < max elements (currently a constant 100).  I don't know how likely we are to hit that limit....

ghalliday commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1134,58 +1161,65 @@ bool EsdlBindingImpl::reloadDefinitionsFromCentralStore(IPropertyTree * esdlBndC
-                tempEsdlBndCfg.set(m_pCentralStore->fetchBinding(process, binding));
+            Owned<IPropertyTree> tempEsdlBndCfg;
+            DBGLOG("Fetching ESDL binding information from dali based on ESP binding (%s)", bindingId);
+            tempEsdlBndCfg.set(m_pCentralStore->fetchBinding(bindingId));
```
It isn't clear from the name fetchBinding whether the pointer it returns in owned or not.  (Normally we use the convention get for linked, and query for not.).   In this case it is, so this call will leak.

ghalliday commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1134,58 +1161,65 @@ bool EsdlBindingImpl::reloadDefinitionsFromCentralStore(IPropertyTree * esdlBndC
-            DBGLOG("Exception caught in EsdlBindingImpl::EsdlBindingImpl: Could Not load Binding %s information from Dali!", binding);
+            m_pESDLService->m_espServiceType.set(loadedname);
+            m_pESDLService->configureTargets(tempEsdlBndCfg, loadedname);
+            m_esdlBndCfg.set(tempEsdlBndCfg.getClear());
```
Leaks.  You either want setown, or 

    m_esdlBndCfg.swap(tempEsdlBndCfg);

ghalliday commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -0,0 +1,574 @@
+    if (gEsdlMonitor.get() != nullptr)
+    {
+        gEsdlMonitor->queryEsdlShare()->stop();
+        gEsdlMonitor.clear();
```
It should also join() to ensure the thread has terminated.

ghalliday reviewed (commented):
As Jake commented, given the large number of memory leaks that have been spotted (in new and old code) this code should be exercised under valgrind, and checking that it doesn't leak.

ghalliday commented:
@mayx some follow on comments.

jakesmith commented:
@mayx - I've added a couple of replies to my previous comments.

mayx commented:
@ghalliday @jakesmith Changes made per latest code review comments. Also fixed several more memory leaks from running Valgrind. Please review again...

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 9ef64179ada9c1a7625588b34702bc802589d7a1
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |784 |784 |0 |
|test (thor) |715 |715 |0 |
|test (roxie) |828 |828 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  169 sec (00:02:49) |   60 sec (00:01:00) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1161 sec (00:19:21) |   19 sec (00:00:19) | 1465 sec (00:24:25) |

jakesmith commented on esp/services/esdl_svc_engine/esdl_monitor.cpp:
```diff
@@ -129,6 +129,7 @@ class CEsdlShare : implements IEsdlShare, public Thread
     {
         m_stop = true;
         m_waitsem.signal();
+        this->join();
```
any reason not the more standard just join(); ?

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 3c2d1f4e4215ef5781497423e375e4b9e4190f55
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |784 |784 |0 |
|test (thor) |715 |715 |0 |
|test (roxie) |828 |828 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  173 sec (00:02:53) |   60 sec (00:01:00) |    8 sec (00:00:08) |   28 sec (00:00:28) | 1156 sec (00:19:16) |   19 sec (00:00:19) | 1459 sec (00:24:19) |

mayx commented:
@jakesmith I've changed this->join() to join() and squashed all changes to one commit. Please do a final review and approve the PR if everything looks OK. (by the way I'll open a separate Jira for changing the bindings array to an expandable array).

jakesmith commented:
ok thanks.
@ghalliday - can you do a final pass ?

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 1647738c280d767a6f2f98c63bce44ce684ce258
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |784 |784 |0 |
|test (thor) |715 |715 |0 |
|test (roxie) |828 |828 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   27 sec (00:00:27) |  169 sec (00:02:49) |   60 sec (00:01:00) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1207 sec (00:20:07) |   18 sec (00:00:18) | 1516 sec (00:25:16) |

ghalliday commented:
@mayx does this fix all of the issues referenced in the commit message, or are some of the fixes only partial?
